File: Trapping_rain_water/code.cpp
class Solution{

    // Function to find the trapped water between the blocks.
    public:
    long long trappingWater(int arr[], int n){
        long long water = 0;
        vector<int> left;
        int maxxi = INT_MIN;
        for(int i=0; i<n; i++){
            maxxi = max(maxxi, arr[i]);
            left.push_back(maxxi);
        }
        vector<int> right;
        maxxi = INT_MIN;
        for(int i=n-1; i>=0; i--){
            maxxi = max(maxxi, arr[i]);
            right.push_back(maxxi);
        }
        reverse(right.begin(), right.end());
        long long maxima;
        for(int i=0; i<n; i++){
            maxima = min(right[i], left[i]) - arr[i];
            if(maxima > 0){
                water+=maxima;
            }
        }
        return water;
    }
};

File: Trapping_rain_water/code.java
class Solution{
    
    // arr: input array
    // n: size of array
    // Function to find the trapped water between the blocks.
    static long trappingWater(int arr[], int n) { 
        //this for leftmax boundary
       int leftmax[]=new int [n];
        leftmax[0]=arr[0];
        for(int i=1;i<n;i++){
            leftmax[i]=Math.max(arr[i],leftmax[i-1]);
        }
        //this for rightmax boundary
       int rightmax[]= new int[n];
        rightmax[n-1]=arr[n-1];
        for(int i=n-2;i>=0;i--){
            rightmax[i]=Math.max(arr[i],rightmax[i+1]);
        }
        //calculate water
        long trappedwater=0;
        for(int i=0;i<n;i++){
            int water=Math.min(leftmax[i],rightmax[i]);
            trappedwater+=water-arr[i];
        }
        return trappedwater;
        
        
    } 
}

